---
title: "snrna-seq_P1_sample1"
author: "Anindita Brahma"
date: "11/11/2022"
output: html_document
---
##10X run from P1_sample1

```{r setup, include=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}
libraries <- c("dplyr", "Seurat",
               "patchwork", "ggplot2")
for (lib in libraries) {
  if (require(package = lib, character.only = TRUE)) {
    message("Library loaded: ", lib)
  } else {
    print("Installing")
    BiocManager::install(lib, Ncpus = 10, update = TRUE)
    library(lib, character.only = TRUE)
    message("Library loaded: ", lib)
  }
}
sessionInfo()
for (package in (.packages())) {
  print(paste("Package", package, "version", packageVersion(package)))
}
```


##Using Seurat on 10X cellranger filtered matrix for P1_sample1

```{r}
#import cellranger count matrix data
p1_sample1 <- Read10X(data.dir = "~/../../Volumes/antqueen/booster/AB20221109_snRNASeq_P1_P14_clubs/count_P1_sample1/outs/filtered_feature_bc_matrix/")

# #check the data
# ##Useful link: https://broadinstitute.github.io/KrumlovSingleCellWorkshop2020/data-wrangling-scrnaseq-1.html
# 
# p1_sample1[1:10, 1:3]
# dim(p1_sample1) ## 14238 genes and 10950 cells
# 
# #summary counts for genes and cells
# p1_sample1_counts_per_cell <- Matrix::colSums(p1_sample1)
# cat("p1_sample1_counts_per_cell: ", p1_sample1_counts_per_cell[1:5], "\n") ##counts for the first 5 cells
# length(p1_sample1_counts_per_cell) 
# 
# p1_sample1_counts_per_gene <- Matrix::rowSums(p1_sample1)
# cat("p1_sample1_counts_per_gene: ", p1_sample1_counts_per_gene[1:5], "\n") ##counts for the first 5 genes
# 
# p1_sample1_genes_per_cell <- Matrix::colSums(p1_sample1 > 0) ##count gene only if it has non-zero reads mapped
# cat("counts for non-zero genes: ", p1_sample1_genes_per_cell[1:5])  ## counts for first 5 genes
# 
# hist(log(p1_sample1_counts_per_cell), main = "counts per cell", col = "wheat")
# hist(log(p1_sample1_genes_per_cell), main = "genes per cell", col = "wheat")
# 
# plot(p1_sample1_counts_per_cell, p1_sample1_genes_per_cell, log = "xy", col = "wheat")
# title("counts vs genes per cell")

# Initialize the Seurat object with the raw (non-normalized data)
p1_sample1_seurat <- CreateSeuratObject(counts = p1_sample1, project = "p1_sample1_antennal_clubs",
                                min.cells = 3, min.features = 200)
p1_sample1_seurat

p1_sample1_seurat@assays$RNA@counts[1:10, 1:10]
dim(p1_sample1_seurat)
str(p1_sample1_seurat)
p1_sample1_seurat[["RNA"]]
GetAssayData(object = p1_sample1_seurat, slot = "data")[1:10, 1:10]


p1_sample1_dense_size <- object.size(as.matrix(x = p1_sample1))
p1_sample1_dense_size

p1_sample1_sparse_size <- object.size(p1_sample1)
p1_sample1_sparse_size

p1_sample1_dense_size/p1_sample1_sparse_size

##QC filtering
# ##mitochondrial contamination
# p1_sample1_seurat[["percent.mt"]] <- PercentageFeatureSet(object = p1_sample1_seurat,
#                                                           pattern = "^MT-")

#Show QC metrics for the first 5 cells
head(p1_sample1_seurat@meta.data, 5)

head(x = p1_sample1_seurat@meta.data[c("nCount_RNA", "nFeature_RNA")])

##visualize QC metrics as a violin plot
VlnPlot(object = p1_sample1_seurat, features = c("nFeature_RNA", "nCount_RNA"),
        ncol = 2)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

#nFeature_RNA = number of genes detected in each cell
#nCount_RNA = total number of molecules detected within a cell

plot1 <- FeatureScatter(object = p1_sample1_seurat, feature1 = "nCount_RNA",
                        feature2 = "nFeature_RNA")
plot1
##plot1 helps to decide the cut-off for filtering cells with very high nFeature_RNA
##In this case it is 4000

#filtering seurat object for low and very high feature counts
p1_sample1_seurat <- subset(x = p1_sample1_seurat,
                           subset = nFeature_RNA > 200 &
                             nFeature_RNA < 4000)

# ##normalize data
# p1_sample1_normalize <- NormalizeData(object = p1_sample1_seurat,
#                                      normalization.method = "LogNormalize",
#                                      scale.factor = 10000)
# #identifying highly variable feautures
# p1_sample1_subset_vf <- FindVariableFeatures(object = p1_sample1_normalize,
#                                      selection.method = "vst",
#                                      nfeatures = 4000)
# # Identify the 10 most highly variable genes
# p1_sample1_vf_top10 <- head(VariableFeatures(object = p1_sample1_subset_vf), 10)
# p1_sample1_vf_top10
# 
# # plot variable features with and without labels
# plot1_sample1_vf <- VariableFeaturePlot(object = p1_sample1_subset_vf)
# plot2_sample1_vf <- LabelPoints(plot = plot1_sample1_vf,
#                         points = p1_sample1_vf_top10)
# plot1_sample1_vf + plot2_sample1_vf

# #scaling the data : linear transformation
# ##The results of this are stored in pbmc[["RNA"]]@scale.data
# p1_sample1_all_genes <- rownames(p1_sample1_subset_vf)
# p1_sample1_subset_vf_scale <- ScaleData(object = p1_sample1_subset_vf,
#                                 features = p1_sample1_all_genes)
# p1_sample1_subset_vf_scale

##use SCTransform instead of normalize data and scale data workflow
p1_sample1_sctransform <- SCTransform(object = p1_sample1_seurat)

##plot normalized gene counts
p1_sample1_normalized_gene_count <- as.vector(x = p1_sample1_sctransform[["SCT"]]@counts)
p1_sample1_normalized_gene_count <- p1_sample1_normalized_gene_count[p1_sample1_normalized_gene_count != 0]
hist(p1_sample1_normalized_gene_count)

##perform linear dimensional reduction
p1_sample1_scale_pca <- RunPCA(object = p1_sample1_sctransform,
                               features = VariableFeatures(object = p1_sample1_sctransform))



# Examine and visualize PCA results a few different ways
print(p1_sample1_scale_pca[["pca"]], dims = 1:5, nfeatures = 5)
VizDimLoadings(object = p1_sample1_scale_pca, dims = 1:2, reduction = "pca")

DimPlot(object = p1_sample1_scale_pca, reduction = "pca")
DimHeatmap(object = p1_sample1_scale_pca, dims = 1, cells = 5000,
           balanced = TRUE)
DimHeatmap(object = p1_sample1_scale_pca, dims = 1:15, cells = 500,
           balanced = TRUE)

##cluster the cells
p1_sample1_scale <- FindNeighbors(object = p1_sample1_scale_pca, dims = 1:10)
p1_sample1_scale <- FindClusters(object = p1_sample1_scale, resolution = 0.5)

##run UMAP/tSNE
p1_sample1_scale <- RunTSNE(object = p1_sample1_scale, dims = 1:10)
DimPlot(object = p1_sample1_scale, reduction = "tsne")
VlnPlot(p1_sample1_scale, features = c("ObirOr5-Orco", "LOC105283685"))
VlnPlot(p1_sample1_scale, features = c("ObirOr5-Orco", "LOC105284919"))



```

##Get the OSNs
```{r}
##get nuclei expressing specific genes
##neuronal markers

#synaptobrevin
p1_sample1_nsyb <- GetAssayData(object = p1_sample1_sctransform,
                     assay = "SCT", slot = "counts")["LOC105284919", ]
head(p1_sample1_nsyb, 10)

#here nuclei identities are rownames; convert that into a column
p1_sample1_nsyb_df <- data.frame(p1_sample1_nsyb)
p1_sample1_nsyb_nuclei_id <- rownames(p1_sample1_nsyb_df)
rownames(p1_sample1_nsyb_df) <- NULL

p1_sample1_nsyb_df <- cbind(p1_sample1_nsyb_nuclei_id, p1_sample1_nsyb_df)
colnames(p1_sample1_nsyb_df) <- c("nuclei_id_nsyb", "nsyb_count")
p1_sample1_nsyb_subset <- subset(x = p1_sample1_nsyb_df, subset = nsyb_count >= 2) ##891

##embryonic lethal abnormal vision
p1_sample1_elav <- GetAssayData(object = p1_sample1_sctransform,
                     assay = "RNA", slot = "data")["LOC105280759", ]
p1_sample1_elav_df <- data.frame(p1_sample1_elav)
p1_sample1_elav_nuclei_id <- rownames(p1_sample1_elav_df)
rownames(p1_sample1_elav_df) <- NULL
p1_sample1_elav_df <- cbind(p1_sample1_elav_nuclei_id, p1_sample1_elav_df)
colnames(p1_sample1_elav_df) <- c("nuclei_id_elav", "elav_count")
p1_sample1_elav_subset <- subset(x = p1_sample1_elav_df, subset = elav_count >= 2) #4418

##brunchpilot
p1_sample1_brp <- GetAssayData(object = p1_sample1_sctransform,
                     assay = "SCT", slot = "counts")["LOC105276401", ]
p1_sample1_brp_df <- data.frame(p1_sample1_brp)
p1_sample1_brp_nuclei_id <- rownames(p1_sample1_brp_df)
rownames(p1_sample1_brp_df) <- NULL
p1_sample1_brp_df <- cbind(p1_sample1_brp_nuclei_id, p1_sample1_brp_df)
colnames(p1_sample1_brp_df) <- c("nuclei_id_brp", "brp_count")
p1_sample1_brp_subset <- subset(x = p1_sample1_brp_df, subset = brp_count >= 2) #7539

##cadherin N
p1_sample1_cadn <- GetAssayData(object = p1_sample1_sctransform,
                     assay = "RNA", slot = "data")["LOC105285306", ]
p1_sample1_cadn_df <- data.frame(p1_sample1_cadn)
p1_sample1_cadn_nuclei_id <- rownames(p1_sample1_cadn_df)
rownames(p1_sample1_cadn_df) <- NULL
p1_sample1_cadn_df <- cbind(p1_sample1_cadn_nuclei_id, p1_sample1_cadn_df)
colnames(p1_sample1_cadn_df) <- c("nuclei_id_cadn", "cadn_count")
p1_sample1_cadn_subset <- subset(x = p1_sample1_cadn_df, subset = cadn_count >= 2) #8029

##synaptotagmin
p1_sample1_syt1 <- GetAssayData(object = p1_sample1_sctransform,
                     assay = "RNA", slot = "data")["LOC105275115", ]
p1_sample1_syt1_df <- data.frame(p1_sample1_syt1)
p1_sample1_syt1_nuclei_id <- rownames(p1_sample1_syt1_df)
rownames(p1_sample1_syt1_df) <- NULL
p1_sample1_syt1_df <- cbind(p1_sample1_syt1_nuclei_id, p1_sample1_syt1_df)
colnames(p1_sample1_syt1_df) <- c("nuclei_id_syt1", "syt1_count")
p1_sample1_syt1_subset <- subset(x = p1_sample1_syt1_df, subset = syt1_count >= 2) #5920

##glial marker
#reverse polarity
p1_sample1_repo <- GetAssayData(object = p1_sample1_sctransform,
                     assay = "RNA", slot = "data")["LOC105286785", ]
p1_sample1_repo_df <- data.frame(p1_sample1_repo)
p1_sample1_repo_nuclei_id <- rownames(p1_sample1_repo_df)
rownames(p1_sample1_repo_df) <- NULL
p1_sample1_repo_df <- cbind(p1_sample1_repo_nuclei_id, p1_sample1_repo_df)
colnames(p1_sample1_repo_df) <- c("nuclei_id_repo", "repo_count")
p1_sample1_repo_subset <- subset(x = p1_sample1_repo_df, subset = repo_count >= 2) ##107


##marker for maturing OSNs (uncharacterized)
p1_sample1_maturing_osn <- GetAssayData(object = p1_sample1_sctransform,
                     assay = "RNA", slot = "data")["LOC105283685", ]
p1_sample1_maturing_osn_df <- data.frame(p1_sample1_maturing_osn)
p1_sample1_maturing_osn_nuclei_id <- rownames(p1_sample1_maturing_osn_df)
rownames(p1_sample1_maturing_osn_df) <- NULL
p1_sample1_maturing_osn_df <- cbind(p1_sample1_maturing_osn_nuclei_id, p1_sample1_maturing_osn_df)
colnames(p1_sample1_maturing_osn_df) <- c("nuclei_id_maturing_osn", "maturing_osn_count")
p1_sample1_maturing_osn_subset <- subset(x = p1_sample1_maturing_osn_df, subset = maturing_osn_count >= 2) ##1788


##get nuclei expressing at least two neuronal markers (corrected UMI count >= 2) 

##nsyb and elav
p1_sample1_nsyb_elav <- data.frame(intersect(p1_sample1_nsyb_subset$nuclei_id_nsyb,
                                             p1_sample1_elav_subset$nuclei_id_elav))
colnames(p1_sample1_nsyb_elav) <- "p1_sample1_nysb_elav_nuclei_ids"

##nysb and brp
p1_sample1_nsyb_brp <- data.frame(intersect(p1_sample1_nsyb_subset$nuclei_id_nsyb,
                                            p1_sample1_brp_subset$nuclei_id_brp))
colnames(p1_sample1_nsyb_brp) <- "p1_sample1_nysb_brp_nuclei_ids"

##nsyb and cadn
p1_sample1_nsyb_cadn <- data.frame(intersect(p1_sample1_nsyb_subset$nuclei_id_nsyb,
                                             p1_sample1_cadn_subset$nuclei_id_cadn))
colnames(p1_sample1_nsyb_cadn) <- "p1_sample1_nysb_cadn_nuclei_ids"

##nsyb and syt1
p1_sample1_nsyb_syt1 <- data.frame(intersect(p1_sample1_nsyb_subset$nuclei_id_nsyb,
                                             p1_sample1_syt1_subset$nuclei_id_syt1))
colnames(p1_sample1_nsyb_syt1) <- "p1_sample1_nysb_syt1_nuclei_ids"

##elav and brp
p1_sample1_elav_brp <- data.frame(intersect(p1_sample1_elav_subset$nuclei_id_elav,
                                            p1_sample1_brp_subset$nuclei_id_brp))
colnames(p1_sample1_elav_brp) <- "p1_sample1_elav_brp_nuclei_ids"

##elav and cadn
p1_sample1_elav_cadn <- data.frame(intersect(p1_sample1_elav_subset$nuclei_id_elav,
                                             p1_sample1_cadn_subset$nuclei_id_cadn))
colnames(p1_sample1_elav_cadn) <- "p1_sample1_elav_cadn_nuclei_ids"

##elav and syt1
p1_sample1_elav_syt1 <- data.frame(intersect(p1_sample1_elav_subset$nuclei_id_elav,
                                             p1_sample1_syt1_subset$nuclei_id_syt1))
colnames(p1_sample1_elav_syt1) <- "p1_sample1_elav_syt1_nuclei_ids"

##brp and cadn
p1_sample1_brp_cadn <- data.frame(intersect(p1_sample1_brp_subset$nuclei_id_brp,
                                            p1_sample1_cadn_subset$nuclei_id_cadn))
colnames(p1_sample1_brp_cadn) <- "p1_sample1_brp_cadn_nuclei_ids"

##brp and syt1
p1_sample1_brp_syt1 <- data.frame(intersect(p1_sample1_brp_subset$nuclei_id_brp,
                                            p1_sample1_syt1_subset$nuclei_id_syt1))
colnames(p1_sample1_brp_syt1) <- "p1_sample1_brp_syt1_nuclei_ids"

##cadn and syt1
p1_sample1_cadn_syt1 <- data.frame(intersect(p1_sample1_cadn_subset$nuclei_id_cadn,
                                             p1_sample1_syt1_subset$nuclei_id_syt1))
colnames(p1_sample1_cadn_syt1) <- "p1_sample1_cadn_syt1_nuclei_ids"

##gather all unique nuclei expressing at least two neuronal markers
p1_sample1_neuronal_uniq_ids <- data.frame(unique(c(p1_sample1_nsyb_elav$p1_sample1_nysb_elav_nuclei_ids,
                                                    p1_sample1_nsyb_brp$p1_sample1_nysb_brp_nuclei_ids,
                                                    p1_sample1_nsyb_cadn$p1_sample1_nysb_cadn_nuclei_ids,
                                                    p1_sample1_nsyb_syt1$p1_sample1_nysb_syt1_nuclei_ids,
                                                    p1_sample1_elav_brp$p1_sample1_elav_brp_nuclei_ids,
                                                    p1_sample1_elav_cadn$p1_sample1_elav_cadn_nuclei_ids,
                                                    p1_sample1_elav_syt1$p1_sample1_elav_syt1_nuclei_ids,
                                                    p1_sample1_brp_cadn$p1_sample1_brp_cadn_nuclei_ids,
                                                    p1_sample1_brp_syt1$p1_sample1_brp_syt1_nuclei_ids,
                                                    p1_sample1_cadn_syt1$p1_sample1_cadn_syt1_nuclei_ids)))
colnames(p1_sample1_neuronal_uniq_ids) <- "p1_sample1_neuronal_uniq_ids" ##7786 nuclei

##subset the normalized seurat object for the barcodes with at least two neuronal markers
p1_sample1_normalize_neuron_ids <- subset(x = p1_sample1_sctransform,
                                  cells = p1_sample1_neuronal_uniq_ids$p1_sample1_neuronal_uniq_ids)
length(colnames(p1_sample1_normalize_neuron_ids))##7786 nuclei
length(rownames(p1_sample1_normalize_neuron_ids)) ##12510 genes

#find the highly variable genes in this subset
p1_sample1_normalize_neuron_ids <- FindVariableFeatures(object = p1_sample1_normalize_neuron_ids,
                                  selection.method = "vst",
                                  nfeatures = 2000)
#identify the top most variable genes
p1_sample1_top20_neuron_genes <- head(VariableFeatures(object = p1_sample1_normalize_neuron_ids), 20)
p1_sample1_plot1 <- VariableFeaturePlot(p1_sample1_normalize_neuron_ids)
p1_sample1_plot2 <- LabelPoints(plot = p1_sample1_plot1, points = p1_sample1_top20_neuron_genes,
                                repel = TRUE)
p1_sample1_plot1 + p1_sample1_plot2
p1_sample1_top20_neuron_genes

# ##scaling the data
# p1_sample1_all_genes_neurons <- rownames(p1_sample1_normalize_neuron_ids)
# 
# ##added vars.to.regress from Herre et al, 2022
# p1_sample1_normalize_neuron_ids <- ScaleData(object = p1_sample1_normalize_neuron_ids,
#                                      features = p1_sample1_all_genes_neurons,
#                                      vars.to.regress = c("nCount_RNA"))

##no need to scale as we have already done sctransform

p1_sample1_normalize_neuron_ids <- RunPCA(object = p1_sample1_normalize_neuron_ids,
                             features = VariableFeatures(object = p1_sample1_normalize_neuron_ids))
DimHeatmap(p1_sample1_normalize_neuron_ids, dims = 1:15, cells = 500, balanced = TRUE)

##cluster the nuclei
p1_sample1_normalize_neuron_ids <- FindNeighbors(object = p1_sample1_normalize_neuron_ids,
                                         dims = 1:10)
p1_sample1_normalize_neuron_ids <- FindClusters(object = p1_sample1_normalize_neuron_ids,
                                        resolution = 0.5)
p1_sample1_normalize_neuron_ids <- RunTSNE(object = p1_sample1_normalize_neuron_ids,
                                   dims = 1:10)
DimPlot(object = p1_sample1_normalize_neuron_ids, reduction = "tsne")

FeaturePlot(object = p1_sample1_normalize_neuron_ids,
            features = c("ObirOr5-Orco"))

##Orco and repo (glial cell marker)
FeaturePlot(object = p1_sample1_normalize_neuron_ids,
            features = c("ObirOr5-Orco", "LOC105286785"))

##Orco and repo and LOC105283685
FeaturePlot(object = p1_sample1_normalize_neuron_ids,
            features = c("ObirOr5-Orco", "LOC105286785", "LOC105283685"))


#cluster markers
#find markers
#cluster 1
p1_sample1_cluster1_markers <- FindMarkers(object = p1_sample1_normalize_neuron_ids, ident.1 = 1,
                                min.pct = 0.25)
head(p1_sample1_cluster1_markers, n = 10)

# gene_count_per_nuclei <- data.frame(p1_sample1_normalize_neuron_ids@assays[["RNA"]]@counts)
# p1_sample1_gene_id <- rownames(gene_count_per_nuclei)
# 
# gene_count_per_nuclei$gene_id <- p1_sample1_gene_id
# rownames(gene_count_per_nuclei) <- NULL
# 
# gene_count_per_nuclei$gene_id <- as.factor(gene_count_per_nuclei$gene_id)
# str(gene_count_per_nuclei$gene_id)
# p1_sample1_or_per_nucleus <- subset(x = gene_count_per_nuclei,
#                          subset = gene_id == list(or_genes)) 
# p1_sample1_or_per_nucleus <- merge(x = gene_count_per_nuclei, y = or_genes)
# ##remove the non-Or genes that are still showing up as factor levels in the df
# p1_sample1_or_per_nucleus <- droplevels(p1_sample1_or_per_nucleus)

##assay set to be SCT due to SCTransform
##counts represents corrected UMI counts that is used for Herre et al 2022
p1_sample1_osn_gene_exp_per_cell <- data.frame(p1_sample1_normalize_neuron_ids@assays[["SCT"]]@counts)

p1_sample1_gene_id <- rownames(p1_sample1_osn_gene_exp_per_cell)

p1_sample1_osn_gene_exp_per_cell$gene_id <- p1_sample1_gene_id
rownames(p1_sample1_osn_gene_exp_per_cell) <- NULL
p1_sample1_osn_gene_exp_per_cell$gene_id <- as.factor(p1_sample1_osn_gene_exp_per_cell$gene_id)
# rownames(p1_sample1_osn_gene_exp_per_cell) <- p1_sample1_osn_gene_exp_per_cell$gene_id
# p1_sample1_osn_gene_exp_per_cell <- select(p1_sample1_osn_gene_exp_per_cell, -gene_id)

##transpose the df
p1_sample1_osn_gene_exp_per_cell_t <- t(p1_sample1_osn_gene_exp_per_cell)

# p1_sample1_or_per_cell <- subset(x = p1_sample1_osn_gene_exp_per_cell,
#                          subset = gene_id == list(or_genes))
p1_sample1_or_per_cell <- merge(x = p1_sample1_osn_gene_exp_per_cell, y = or_genes)
##remove the non-Or genes that are still showing up as factor levels in the df
p1_sample1_or_per_cell <- droplevels(p1_sample1_or_per_cell)
rownames(p1_sample1_or_per_cell) <- p1_sample1_or_per_cell$gene_id
p1_sample1_or_per_cell <- select(p1_sample1_or_per_cell, -gene_id)

# or_uncharc <- data.frame(as.factor(c("ObirOr5-Orco", "LOC105283685")))
# colnames(or_uncharc) <- "gene_id"
# 
# p1_sample1_or_uncharc <- merge(x = p1_sample1_osn_gene_exp_per_cell, y = or_uncharc)




##calculate number of ORs per nuclei with UMI count >= 2

#transpose the data such that gene ids are now the columns
p1_sample1_or_per_cell_t <- data.frame(t(p1_sample1_or_per_cell))
# p1_sample1_or_per_cell_t$cell_barcode <- as.factor(rownames(p1_sample1_or_per_cell_t))
# rownames(p1_sample1_or_per_cell_t) <- NULL
# or_per_nucleus_t$nuclei_barcodes <- as.factor(nuclei_barcodes)

length(p1_sample1_or_per_cell_t$ObirOr5.9E1) ##this should be 7786 (# of neuronal cells)
#this number is off because gene_id is added as a row at the end while transposing
#remove the last row i.e., row # 6656 here
p1_sample1_or_per_cell_t <- p1_sample1_or_per_cell_t[-7787, ]
length(p1_sample1_or_per_cell_t$`ObirOr5.9E1`) ##now this is 7786

# colnames(p1_sample1_or_per_cell_t) <- p1_sample1_or_per_cell$gene_id
p1_sample1_cell_barcodes <- rownames(p1_sample1_or_per_cell_t)

# ##get cells with Orco expression >= 2
p1_sample1_orco <- subset(x = p1_sample1_or_per_cell_t, subset = ObirOr5.Orco >= 2)
p1_sample1_orco_t <- t(p1_sample1_orco)
gene_id_orco_cells <- data.frame(as.factor(rownames(p1_sample1_orco_t)))
colnames(gene_id_orco_cells) <- "gene_id_orco_cells"
rownames(p1_sample1_orco_t) <- NULL

write.table(x = p1_sample1_orco_t, file = "results/p1_sample1_orco_cells.csv",
            sep = ",", row.names = FALSE)
write.table(x = gene_id_orco_cells, file = "results/p1_sample1_gene_id_orco_cells.csv",
            sep = ",", row.names = FALSE)

p1_sample1_orco_t$gene_id <- gene_id_orco_cells$gene_id_orco_cells

# p1_sample1_uncharc <- subset(x = p1_sample1_osn_gene_exp_per_cell_t, subset = ObirOr5.Orco >= 2)

##create a vector with unique cell barcodes (there shouldn't be duplication though)
##confirmed that there is no duplication
p1_sample1_cell_barcode_vec <- unique(rownames(p1_sample1_orco))
length(p1_sample1_cell_barcode_vec) 


#create empty dataframe
p1_sample1_result_df <- as.data.frame(matrix(ncol = 3,
                                  nrow = length(p1_sample1_cell_barcode_vec)))
colnames(p1_sample1_result_df) <- c("cell_barcode", "false", "true")
p1_sample1_result_df$cell_barcode <- p1_sample1_cell_barcode_vec

for (position in seq_len(length(p1_sample1_cell_barcode_vec))) {
  print(position)
  ###only select those ORs with lognorm >2 for a particular cell
  p1_sample1_table <- data.frame(table(p1_sample1_or_per_cell_t[position, ] > 1 )) 
  p1_sample1_table_t <- data.frame(t(p1_sample1_table))
  
  if (colnames(p1_sample1_table_t) == "t.p1_sample1_table.") {
    p1_sample1_result_df$false[position] <- 548
    p1_sample1_result_df$true[position] <- 0
  }
  
  if (colnames(p1_sample1_table_t)[1] == "X1") {
    colnames(p1_sample1_table_t) <- c("false", "true")
  p1_sample1_table_t <- p1_sample1_table_t[-1, ]
  p1_sample1_result_df$false[position] <- p1_sample1_table_t$false
  p1_sample1_result_df$true[position] <- p1_sample1_table_t$true
  }
  
} 

#true column in this table represents number of ORs with lognormal expression > 2
#save table
# write.table(x = p1_sample1_result_df, file = "results/p1_sample1_or_per_cell.csv",
#             sep = ",", quote = FALSE, row.names = FALSE)

#remove rownames
rownames(p1_sample1_result_df) <- NULL

#remove whitespace from the column true
p1_sample1_result_df_copy <- data.frame(gsub(pattern = " ", replacement = "",
                       x = p1_sample1_result_df$true))
colnames(p1_sample1_result_df_copy) <- "true_or"
str(p1_sample1_result_df_copy$true_or)

#change from character to numeric
p1_sample1_result_df_copy$true_or <- as.numeric(p1_sample1_result_df_copy$true_or)

#add this into the original result_df dataframe
p1_sample1_result_df$true_or <- gsub(pattern = " ", replacement = "",
                                     x = p1_sample1_result_df$true)
str(p1_sample1_result_df)
p1_sample1_result_df$true_or <- as.numeric(p1_sample1_result_df$true_or)

##subset the nuclei for which there is 1 OR with read count > 1
p1_sample1_only_one_or <- subset(x = p1_sample1_result_df, subset = true_or == 1) ##143

##subset the nuclei for which there are more than 1 ORs with read count > 1
p1_sample1_more_than_one_or <- subset(x = p1_sample1_result_df, subset = true_or > 1) #281

#subset the nuclei for which there are more than or = 2 and less than 4 ORs with read count > 1
p1_sample1_more_than_two_ors <- subset(x = p1_sample1_result_df, subset = true_or >= 2 &
                              true_or < 4) #118
#subset for more than 4 ORs
p1_sample1_more_than_four_ors <- subset(x = p1_sample1_result_df, subset = true_or >= 4) ##163

#subset for nuclei with no ORs with read count > 1
p1_sample1_no_or <- subset(x = p1_sample1_result_df, subset = true_or == 0) ##1

#make a freq distribution
p1_sample1_categories <- c("no_OR", "only_one_OR",
                "two_or_less_than_four", "four_or_more_ORs")
p1_sample1_values <- c(1, 7, 21, 192)

p1_sample1_hist_df <- data.frame(p1_sample1_categories, p1_sample1_values)
p1_sample1_hist_df_t <- data.frame(t(p1_sample1_hist_df))

barplot((p1_sample1_hist_df$p1_sample1_values))

barplot_df <- data.frame(name = p1_sample1_categories,
                         value = p1_sample1_values)
barplot_p1_sample1 <- ggplot(data = barplot_df, aes(x = name, y = value)) +
  geom_bar(stat = "identity") +
  xlab("Number of ORs in neurons with Orco expression") +
  ylab("Frequency") +
  theme(axis.text.x = element_text(angle = 90))




```



```{r}
##list of OR genes
or_genes <- c("ObirOr5-9E1",
"ObirOr5-9E10",
"ObirOr5-9E100",
"ObirOr5-9E101",
"ObirOr5-9E102",
"ObirOr5-9E103",
"ObirOr5-9E104",
"ObirOr5-9E105PSE",
"ObirOr5-9E106",
"ObirOr5-9E107",
"ObirOr5-9E108",
"ObirOr5-9E109",
"ObirOr5-9E11",
"ObirOr5-9E110",
"ObirOr5-9E111",
"ObirOr5-9E112",
"ObirOr5-9E113",
"ObirOr5-9E114",
"ObirOr5-9E115",
"ObirOr5-9E116",
"ObirOr5-9E117",
"ObirOr5-9E118",
"ObirOr5-9E119",
"ObirOr5-9E12",
"ObirOr5-9E120",
"ObirOr5-9E121",
"ObirOr5-9E122",
"ObirOr5-9E123PSE",
"ObirOr5-9E124",
"ObirOr5-9E125",
"ObirOr5-9E126",
"ObirOr5-9E127",
"ObirOr5-9E128",
"ObirOr5-9E129",
"ObirOr5-9E13",
"ObirOr5-9E130PSE",
"ObirOr5-9E131",
"ObirOr5-9E132",
"ObirOr5-9E133",
"ObirOr5-9E134",
"ObirOr5-9E135CTE",
"ObirOr5-9E136",
"ObirOr5-9E137",
"ObirOr5-9E138",
"ObirOr5-9E139",
"ObirOr5-9E14",
"ObirOr5-9E140",
"ObirOr5-9E141",
"ObirOr5-9E142PSE",
"ObirOr5-9E143PSE",
"ObirOr5-9E144PSE",
"ObirOr5-9E145PSE",
"ObirOr5-9E146PSE",
"ObirOr5-9E147PSE",
"ObirOr5-9E148",
"ObirOr5-9E149PSE",
"ObirOr5-9E15",
"ObirOr5-9E150",
"ObirOr5-9E151",
"ObirOr5-9E152",
"ObirOr5-9E153PSE",
"ObirOr5-9E154",
"ObirOr5-9E155",
"ObirOr5-9E156",
"ObirOr5-9E157",
"ObirOr5-9E158",
"ObirOr5-9E159",
"ObirOr5-9E16",
"ObirOr5-9E160",
"ObirOr5-9E161",
"ObirOr5-9E162",
"ObirOr5-9E163PSE",
"ObirOr5-9E164",
"ObirOr5-9E165",
"ObirOr5-9E166",
"ObirOr5-9E167",
"ObirOr5-9E168",
"ObirOr5-9E169",
"ObirOr5-9E17",
"ObirOr5-9E170",
"ObirOr5-9E171",
"ObirOr5-9E172",
"ObirOr5-9E173",
"ObirOr5-9E174",
"ObirOr5-9E175",
"ObirOr5-9E176",
"ObirOr5-9E177",
"ObirOr5-9E178PSE",
"ObirOr5-9E179",
"ObirOr5-9E18",
"ObirOr5-9E180",
"ObirOr5-9E181",
"ObirOr5-9E182",
"ObirOr5-9E183PSE",
"ObirOr5-9E184",
"ObirOr5-9E185",
"ObirOr5-9E186",
"ObirOr5-9E187PSE",
"ObirOr5-9E188",
"ObirOr5-9E189",
"ObirOr5-9E19",
"ObirOr5-9E190",
"ObirOr5-9E191",
"ObirOr5-9E192",
"ObirOr5-9E193",
"ObirOr5-9E194",
"ObirOr5-9E195",
"ObirOr5-9E196",
"ObirOr5-9E197",
"ObirOr5-9E198",
"ObirOr5-9E199PSE",
"ObirOr5-9E20",
"ObirOr5-9E200",
"ObirOr5-9E201",
"ObirOr5-9E202",
"ObirOr5-9E203",
"ObirOr5-9E204",
"ObirOr5-9E205",
"ObirOr5-9E206",
"ObirOr5-9E207",
"ObirOr5-9E208",
"ObirOr5-9E209",
"ObirOr5-9E21",
"ObirOr5-9E210",
"ObirOr5-9E211",
"ObirOr5-9E212",
"ObirOr5-9E213",
"ObirOr5-9E214",
"ObirOr5-9E215",
"ObirOr5-9E216",
"ObirOr5-9E217",
"ObirOr5-9E218",
"ObirOr5-9E219",
"ObirOr5-9E22",
"ObirOr5-9E220",
"ObirOr5-9E221",
"ObirOr5-9E222",
"ObirOr5-9E223",
"ObirOr5-9E224",
"ObirOr5-9E225",
"ObirOr5-9E226",
"ObirOr5-9E227",
"ObirOr5-9E228",
"ObirOr5-9E229",
"ObirOr5-9E23",
"ObirOr5-9E230",
"ObirOr5-9E231",
"ObirOr5-9E232",
"ObirOr5-9E233",
"ObirOr5-9E234PSE",
"ObirOr5-9E235PSE",
"ObirOr5-9E236PSE",
"ObirOr5-9E237PSE",
"ObirOr5-9E238",
"ObirOr5-9E239",
"ObirOr5-9E240PSE",
"ObirOr5-9E241",
"ObirOr5-9E242",
"ObirOr5-9E243",
"ObirOr5-9E244",
"ObirOr5-9E245",
"ObirOr5-9E246",
"ObirOr5-9E247",
"ObirOr5-9E248",
"ObirOr5-9E249",
"ObirOr5-9E24PSE",
"ObirOr5-9E25",
"ObirOr5-9E250",
"ObirOr5-9E251",
"ObirOr5-9E252PSE",
"ObirOr5-9E253PSE",
"ObirOr5-9E254",
"ObirOr5-9E255",
"ObirOr5-9E256",
"ObirOr5-9E257",
"ObirOr5-9E258",
"ObirOr5-9E259",
"ObirOr5-9E26",
"ObirOr5-9E260",
"ObirOr5-9E261",
"ObirOr5-9E262PSE",
"ObirOr5-9E263",
"ObirOr5-9E264PSE",
"ObirOr5-9E265PSE",
"ObirOr5-9E266PSE",
"ObirOr5-9E267",
"ObirOr5-9E268",
"ObirOr5-9E269",
"ObirOr5-9E27",
"ObirOr5-9E270",
"ObirOr5-9E271",
"ObirOr5-9E272",
"ObirOr5-9E273",
"ObirOr5-9E274",
"ObirOr5-9E275",
"ObirOr5-9E276",
"ObirOr5-9E277",
"ObirOr5-9E278",
"ObirOr5-9E279",
"ObirOr5-9E28",
"ObirOr5-9E280",
"ObirOr5-9E281",
"ObirOr5-9E282PSE",
"ObirOr5-9E283",
"ObirOr5-9E284",
"ObirOr5-9E285",
"ObirOr5-9E286",
"ObirOr5-9E287",
"ObirOr5-9E288",
"ObirOr5-9E289",
"ObirOr5-9E29",
"ObirOr5-9E290",
"ObirOr5-9E291",
"ObirOr5-9E292",
"ObirOr5-9E293",
"ObirOr5-9E294",
"ObirOr5-9E295",
"ObirOr5-9E296",
"ObirOr5-9E297",
"ObirOr5-9E298",
"ObirOr5-9E299",
"ObirOr5-9E2PSE",
"ObirOr5-9E30",
"ObirOr5-9E300",
"ObirOr5-9E301",
"ObirOr5-9E302",
"ObirOr5-9E303",
"ObirOr5-9E304",
"ObirOr5-9E305",
"ObirOr5-9E306",
"ObirOr5-9E307",
"ObirOr5-9E308",
"ObirOr5-9E309",
"ObirOr5-9E31",
"ObirOr5-9E310",
"ObirOr5-9E311",
"ObirOr5-9E312",
"ObirOr5-9E313",
"ObirOr5-9E314PSE",
"ObirOr5-9E315",
"ObirOr5-9E316",
"ObirOr5-9E317",
"ObirOr5-9E318",
"ObirOr5-9E319",
"ObirOr5-9E32",
"ObirOr5-9E320",
"ObirOr5-9E321",
"ObirOr5-9E322",
"ObirOr5-9E323",
"ObirOr5-9E324",
"ObirOr5-9E325",
"ObirOr5-9E326",
"ObirOr5-9E327",
"ObirOr5-9E328",
"ObirOr5-9E329",
"ObirOr5-9E33",
"ObirOr5-9E330",
"ObirOr5-9E331",
"ObirOr5-9E332",
"ObirOr5-9E333",
"ObirOr5-9E334",
"ObirOr5-9E335",
"ObirOr5-9E336",
"ObirOr5-9E337",
"ObirOr5-9E338",
"ObirOr5-9E339",
"ObirOr5-9E34",
"ObirOr5-9E340",
"ObirOr5-9E341PSE",
"ObirOr5-9E342",
"ObirOr5-9E343",
"ObirOr5-9E344",
"ObirOr5-9E345",
"ObirOr5-9E346",
"ObirOr5-9E347",
"ObirOr5-9E348",
"ObirOr5-9E349",
"ObirOr5-9E35",
"ObirOr5-9E350",
"ObirOr5-9E351",
"ObirOr5-9E352",
"ObirOr5-9E353",
"ObirOr5-9E354PSE",
"ObirOr5-9E36",
"ObirOr5-9E37",
"ObirOr5-9E38",
"ObirOr5-9E39PSE",
"ObirOr5-9E3PSE",
"ObirOr5-9E4",
"ObirOr5-9E40PSE",
"ObirOr5-9E41",
"ObirOr5-9E42",
"ObirOr5-9E43",
"ObirOr5-9E44",
"ObirOr5-9E45",
"ObirOr5-9E46",
"ObirOr5-9E47",
"ObirOr5-9E48",
"ObirOr5-9E49",
"ObirOr5-9E5",
"ObirOr5-9E50",
"ObirOr5-9E51",
"ObirOr5-9E52",
"ObirOr5-9E53PSE",
"ObirOr5-9E54",
"ObirOr5-9E55",
"ObirOr5-9E56",
"ObirOr5-9E57",
"ObirOr5-9E58",
"ObirOr5-9E59",
"ObirOr5-9E6",
"ObirOr5-9E60",
"ObirOr5-9E61",
"ObirOr5-9E62",
"ObirOr5-9E63PSE",
"ObirOr5-9E64",
"ObirOr5-9E65",
"ObirOr5-9E66",
"ObirOr5-9E67",
"ObirOr5-9E68",
"ObirOr5-9E69",
"ObirOr5-9E7",
"ObirOr5-9E70",
"ObirOr5-9E71",
"ObirOr5-9E72PSE",
"ObirOr5-9E73",
"ObirOr5-9E74",
"ObirOr5-9E75",
"ObirOr5-9E76",
"ObirOr5-9E77",
"ObirOr5-9E78",
"ObirOr5-9E79",
"ObirOr5-9E8",
"ObirOr5-9E80",
"ObirOr5-9E81",
"ObirOr5-9E82",
"ObirOr5-9E83",
"ObirOr5-9E84",
"ObirOr5-9E85",
"ObirOr5-9E86",
"ObirOr5-9E87",
"ObirOr5-9E88",
"ObirOr5-9E89",
"ObirOr5-9E9",
"ObirOr5-9E90",
"ObirOr5-9E91",
"ObirOr5-9E92",
"ObirOr5-9E93",
"ObirOr5-9E94",
"ObirOr5-9E95",
"ObirOr5-9E96",
"ObirOr5-9E97",
"ObirOr5-9E98",
"ObirOr5-9E99",
"ObirOr5-9Efrag1",
"ObirOr5-9Efrag2",
"ObirOr5-9Efrag3",
"ObirOr5-A1",
"ObirOr5-A2",
"ObirOr5-A3",
"ObirOr5-A4",
"ObirOr5-A5",
"ObirOr5-A6",
"ObirOr5-A7",
"ObirOr5-B1",
"ObirOr5-C1",
"ObirOr5-D1",
"ObirOr5-D2",
"ObirOr5-E10",
"ObirOr5-E11",
"ObirOr5-E12",
"ObirOr5-E13",
"ObirOr5-E14",
"ObirOr5-E15",
"ObirOr5-E16",
"ObirOr5-E17",
"ObirOr5-E18",
"ObirOr5-E19",
"ObirOr5-E1PSE",
"ObirOr5-E20",
"ObirOr5-E21",
"ObirOr5-E22",
"ObirOr5-E23",
"ObirOr5-E24",
"ObirOr5-E25",
"ObirOr5-E26",
"ObirOr5-E2PSE",
"ObirOr5-E3",
"ObirOr5-E4",
"ObirOr5-E5",
"ObirOr5-E6",
"ObirOr5-E7",
"ObirOr5-E8",
"ObirOr5-E9",
"ObirOr5-F1",
"ObirOr5-G1",
"ObirOr5-H1",
"ObirOr5-H2",
"ObirOr5-H3",
"ObirOr5-H4PSE",
"ObirOr5-H5",
"ObirOr5-H6PSE",
"ObirOr5-Hfrag1",
"ObirOr5-I1",
"ObirOr5-J1",
"ObirOr5-J2PSE",
"ObirOr5-J3",
"ObirOr5-K1",
"ObirOr5-K2",
"ObirOr5-L10",
"ObirOr5-L11",
"ObirOr5-L12",
"ObirOr5-L13",
"ObirOr5-L14",
"ObirOr5-L15",
"ObirOr5-L16",
"ObirOr5-L17PSE",
"ObirOr5-L18",
"ObirOr5-L19PSE",
"ObirOr5-L1PSE",
"ObirOr5-L2",
"ObirOr5-L20",
"ObirOr5-L21PSE",
"ObirOr5-L22",
"ObirOr5-L23",
"ObirOr5-L24PSE",
"ObirOr5-L25",
"ObirOr5-L26",
"ObirOr5-L27",
"ObirOr5-L28",
"ObirOr5-L29",
"ObirOr5-L3",
"ObirOr5-L30",
"ObirOr5-L31",
"ObirOr5-L32",
"ObirOr5-L33",
"ObirOr5-L34",
"ObirOr5-L35",
"ObirOr5-L4",
"ObirOr5-L5",
"ObirOr5-L6",
"ObirOr5-L7",
"ObirOr5-L8",
"ObirOr5-L9",
"ObirOr5-M1",
"ObirOr5-M2",
"ObirOr5-N1",
"ObirOr5-N2",
"ObirOr5-N3",
"ObirOr5-N4PSE",
"ObirOr5-O1",
"ObirOr5-O2",
"ObirOr5-O3",
"ObirOr5-Orco",
"ObirOr5-P1",
"ObirOr5-P2",
"ObirOr5-P3PSE",
"ObirOr5-P4",
"ObirOr5-P5",
"ObirOr5-P6",
"ObirOr5-P7",
"ObirOr5-Q1",
"ObirOr5-Q2",
"ObirOr5-Q3",
"ObirOr5-R1PSE",
"ObirOr5-R2",
"ObirOr5-R3",
"ObirOr5-R4PSE",
"ObirOr5-S1PSE",
"ObirOr5-T1",
"ObirOr5-T2",
"ObirOr5-T3",
"ObirOr5-U1",
"ObirOr5-U10",
"ObirOr5-U11",
"ObirOr5-U12",
"ObirOr5-U13",
"ObirOr5-U14",
"ObirOr5-U15",
"ObirOr5-U16",
"ObirOr5-U17",
"ObirOr5-U18",
"ObirOr5-U19PSE",
"ObirOr5-U2",
"ObirOr5-U20",
"ObirOr5-U21",
"ObirOr5-U22",
"ObirOr5-U23",
"ObirOr5-U24",
"ObirOr5-U25",
"ObirOr5-U26PSE",
"ObirOr5-U27",
"ObirOr5-U28",
"ObirOr5-U29",
"ObirOr5-U3",
"ObirOr5-U30PSE",
"ObirOr5-U31",
"ObirOr5-U32",
"ObirOr5-U33",
"ObirOr5-U34",
"ObirOr5-U35",
"ObirOr5-U36",
"ObirOr5-U37",
"ObirOr5-U38",
"ObirOr5-U39",
"ObirOr5-U4",
"ObirOr5-U40",
"ObirOr5-U41PSE",
"ObirOr5-U42",
"ObirOr5-U43",
"ObirOr5-U44",
"ObirOr5-U45",
"ObirOr5-U46PSE",
"ObirOr5-U47",
"ObirOr5-U48",
"ObirOr5-U49PSE",
"ObirOr5-U5",
"ObirOr5-U50",
"ObirOr5-U51",
"ObirOr5-U52",
"ObirOr5-U53HET",
"ObirOr5-U54",
"ObirOr5-U55",
"ObirOr5-U56",
"ObirOr5-U57",
"ObirOr5-U58",
"ObirOr5-U59",
"ObirOr5-U6",
"ObirOr5-U60HET",
"ObirOr5-U61",
"ObirOr5-U62",
"ObirOr5-U63",
"ObirOr5-U64",
"ObirOr5-U65",
"ObirOr5-U67",
"ObirOr5-U7",
"ObirOr5-U8",
"ObirOr5-U9",
"ObirOr5-Ufrag1",
"ObirOr5-V1",
"ObirOr5-V10",
"ObirOr5-V11",
"ObirOr5-V12PSE",
"ObirOr5-V13",
"ObirOr5-V14",
"ObirOr5-V15",
"ObirOr5-V16",
"ObirOr5-V17",
"ObirOr5-V18",
"ObirOr5-V19PSE",
"ObirOr5-V20PSE",
"ObirOr5-V21",
"ObirOr5-V22",
"ObirOr5-V23",
"ObirOr5-V24",
"ObirOr5-V25",
"ObirOr5-V26",
"ObirOr5-V27",
"ObirOr5-V28",
"ObirOr5-V2PSE",
"ObirOr5-V3",
"ObirOr5-V4",
"ObirOr5-V5",
"ObirOr5-V6",
"ObirOr5-V7",
"ObirOr5-V8",
"ObirOr5-V9",
"ObirOr5-W1",
"ObirOr5-X1",
"ObirOr5-X2PSE",
"ObirOr5-X3PSE",
"ObirOr5-XA1",
"ObirOr5-Y1",
"ObirOr5-ZA1")

or_genes <- data.frame(or_genes)
colnames(or_genes) <- "gene_id"
str(or_genes)
or_genes$or_genes <- as.factor(or_genes$gene_id)

```

